### PR TITLE
[RoosterTeeth] Add new extractor

### PR DESCRIPTION
**Pull request by nehalvpatel**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added a new extractor for all [Rooster Teeth](https://www.roosterteeth.com/) sites. In addition to the video, it extracts the description and the thumbnail. It can download both public and sponsor-only videos (login).

### CODE REVIEW

**Comment by nehalvpatel** [comment]
> m3u8 download detected but ffmpeg or avconv could not be found. Please install one.

Seems like there was an error with ffmpeg? It extracted the m3u8 URL just fine.

File: youtube_dl/extractor/roosterteeth.py [hunk 1]
```diff
@@ -81,6 +84,6 @@ def _real_extract(self, url):
             'title': title,
             'thumbnail': self._og_search_thumbnail(webpage),
             'description': self._og_search_description(webpage),
-            'creator': self._og_search_property("site_name", webpage, "site"),
+            'creator': self._html_search_regex(r'body id="episode" class="site-([^"]*)"', webpage, "site", fatal=False),
```
Comment (TRox1972) [inline youtube_dl/extractor/roosterteeth.py]: "site" should be in single quotes

Comment (nehalvpatel) [inline youtube_dl/extractor/roosterteeth.py]: Fixed in e8b6c00567ad716d6efa558a112b25307a801720

File: youtube_dl/extractor/roosterteeth.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    sanitized_Request,
+    urlencode_postdata,
+)
+
+
+class RoosterTeethIE(InfoExtractor):
+    _VALID_URL = r'https?:\/\/(?:(.+)\.)?roosterteeth\.com\/episode\/(?P<id>[^\r\n]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: No need to escape `/`.

File: youtube_dl/extractor/roosterteeth.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    def _login(self):
+        (username, password) = self._get_login_info()
+        if username is None or password is None:
+            return False
+
+        # token is required to authenticate request
+        login_page = self._download_webpage(self._LOGIN_URL, None, 'Getting login token', 'Unable to get login token')
+        token = self._html_search_regex(r'name="_token" type="hidden" value="([^"]*)"', login_page, 'login token')
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Use `self._hidden_inputs` instead.

File: youtube_dl/extractor/roosterteeth.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            '_token': token,
+            'username': username,
+            'password': password,
+        }
+
+        payload = urlencode_postdata(form_data)
+
+        request = sanitized_Request(self._LOGIN_URL, payload)
+        request.add_header('Referer', 'https://roosterteeth.com/login')
+        request.add_header('Content-Type', 'application/x-www-form-urlencoded')
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: 41-51 can be passed directly to `_download_webpage`.

File: youtube_dl/extractor/roosterteeth.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+        self.report_login()
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: As well as this.

File: youtube_dl/extractor/roosterteeth.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                'Login failed (invalid username/password)', expected=True)
+
+    def _real_initialize(self):
+        self._login()
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        self.report_extraction(video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Superfluous message.

Comment (nehalvpatel) [inline youtube_dl/extractor/roosterteeth.py]: @dstftw Fixed in a22bbb1adffda4b1c04b01a7983c759544ca76ca

File: youtube_dl/extractor/roosterteeth.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    urlencode_postdata,
+)
+
+
+class RoosterTeethIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:(.+)\.)?roosterteeth\.com/episode/(?P<id>[^/?#&]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Do not capture group if you don't use it. Add `only_matching` tests for subdomains. Subdomain regex should not be greedy.

**Comment by dstftw** [comment]
Squash commits when done.

**Comment by nehalvpatel** [comment]
@dstftw Relaxed the regex and squashed the commits: 65855424d3117d971b087935d18bf77b1b0ee0b1

**Comment by dstftw** [comment]
Not actually. You've restricted it. And still capture the group you don't use.

File: youtube_dl/extractor/roosterteeth.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _login(self):
+        (username, password) = self._get_login_info()
+        if username is None or password is None:
+            return False
+
+        # token is required to authenticate request
+        login_page = self._download_webpage(self._LOGIN_URL, None, 'Getting login token', 'Unable to get login token')
+        hidden_inputs = self._hidden_inputs(login_page)
+        token = hidden_inputs['_token']
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: I've already suggested using this whole dictionary. `_token` is already extracted you don't need to access it explicitly.

File: youtube_dl/extractor/roosterteeth.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            '_token': token,
+            'username': username,
+            'password': password,
+        }
+
+        login_payload = urlencode_postdata(login_data)
+
+        # required for proper responses
+        login_headers = {
+            'Referer': 'https://roosterteeth.com/login',
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Code duplication. Replace with `self._LOGIN_URL`.

File: youtube_dl/extractor/roosterteeth.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        login_request = self._download_webpage(self._LOGIN_URL, None, note='Logging in as %s' % username, data=login_payload, headers=login_headers)
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Carry long lines.

File: youtube_dl/extractor/roosterteeth.py (hunk 8 above)
Comment (nehalvpatel) [inline youtube_dl/extractor/roosterteeth.py]: To be honest, I'm new to Python and I didn't know what you meant. Please explain.

Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Add `username` and `password` to `hidden_inputs` and use it as payload.

**Comment by nehalvpatel** [comment]
@dstftw Right now, it's not capturing the subdomain group, just matching it. The `?:` prevents the capture. Besides, how else can it detect all the possible subdomains?

**Comment by dstftw** [comment]
Wrong. You have inner group that is captured `(www|achievementhunter|funhaus|screwattack|theknow)`.

**Comment by nehalvpatel** [comment]
If you take a look [here](https://regex101.com/r/tH2tR2/1), only the episode id is matched. Unless I'm misunderstood, the `?:` marks the inner group as non-capturing.

File: youtube_dl/extractor/roosterteeth.py (hunk 8 above)
Comment (nehalvpatel) [inline youtube_dl/extractor/roosterteeth.py]: Thanks, will do.

**Comment by dstftw** [comment]
If you take a look at the string you'are matching you may notice that you have nothing to capture by your first group. Change it to `http://www.roosterteeth.com/episode/day-5-day-5-episode-1-waking-nightmare` and see.

**Comment by nehalvpatel** [comment]
Oh okay I see.

How about this: https://regex101.com/r/tH2tR2/2

**Comment by dstftw** [comment]
Just change to ``(?:[\da-zA-Z-]+\.)?roosterteeth\.com` or `(?:.+?\.)?roosterteeth\.com`.

**Comment by nehalvpatel** [comment]
Thanks, changed it.

**Pull request closed without merging**